### PR TITLE
[SPARK-31467][SQL][TEST] Refactor the sql tests to prevent TableAlreadyExistsException

### PR DESCRIPTION
**Pull request by turboFei**
### What changes were proposed in this pull request?
If we add UT in hive/SQLQuerySuite or other sql test suites and use table named `test`.
We may meet TableAlreadyExistsException.

```
org.apache.spark.sql.catalyst.analysis.TableAlreadyExistsException: Table or view 'test' already exists in database 'default'
```

The reason is that, there is some tests that does not clean up the tables/views.
In this PR, I add `withTempViews` for these tests.

### Why are the changes needed?
To fix the TableAlreadyExistsException issue when adding an UT, which uses table named `test` or others, in some sql test suites, such as hive/SQLQuerySuite.

### Does this PR introduce any user-facing change?
No.

### How was this patch tested?

Existed UT.

### CODE REVIEW

**Comment by turboFei** [comment]
It seems that there are many tests do not clean up used tables/views.
I will try to refactor this suite.

**Comment by turboFei** [comment]
Just add withTempView for the tests that do not clean up used tables/views.
cc @HyukjinKwon

**Comment by maropu** [comment]
ok to test

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/SQLQuerySuite.scala [hunk 1]
```diff
@@ -898,33 +913,39 @@ class SQLQuerySuite extends QueryTest with SQLTestUtils with TestHiveSingleton {
...
   test("resolve udtf in projection #1") {
-    val ds = (1 to 5).map(i => s"""{"a":[$i, ${i + 1}]}""").toDS()
-    read.json(ds).createOrReplaceTempView("data")
-    val df = sql("SELECT explode(a) AS val FROM data")
-    val col = df("val")
+    withTempView("data") {
+      val ds = (1 to 5).map(i => s"""{"a":[$i, ${i + 1}]}""").toDS()
+      read.json(ds).createOrReplaceTempView("data")
+      val df = sql("SELECT explode(a) AS val FROM data")
+      val col = df("val")
```
Comment (maropu) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/SQLQuerySuite.scala:920]: (Not related to this pr though), `col` not used?

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/SQLQuerySuite.scala [hunk 2]
```diff
@@ -974,47 +995,57 @@ class SQLQuerySuite extends QueryTest with SQLTestUtils with TestHiveSingleton {
...
-      sql("""FROM
-          |(FROM test SELECT TRANSFORM(key, value) USING 'cat' AS (`thing1` int, thing2 string)) t
-          |SELECT thing1 + 1
+    withTempView("test") {
+      assume(TestUtils.testCommandAvailable("/bin/bash"))
+      val data = (1 to 5).map { i => (i, i) }
+      data.toDF("key", "value").createOrReplaceTempView("test")
+      checkAnswer(
+        sql("""FROM
+              |(FROM test SELECT TRANSFORM(key, value) USING 'cat' AS (`thing1` int, thing2 string))
```
Comment (maropu) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/SQLQuerySuite.scala:1045]: nit: wrong indents.

**Comment by maropu** [comment]
I think the change looks reasonable. Did you check them only in `SQLQuerySuite`? I mean, have you checked if the logs of the other test suites have no `TableAlreadyExistsException`?

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/SQLQuerySuite.scala (hunk 1 above)
Comment (turboFei) [reply sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/SQLQuerySuite.scala:920]: yes, should we remove this line?

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/SQLQuerySuite.scala (hunk 2 above)
Comment (turboFei) [reply sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/SQLQuerySuite.scala:1045]: thanks

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/SQLQuerySuite.scala (hunk 1 above)
Comment (turboFei) [reply sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/SQLQuerySuite.scala:920]: removed

**Comment by turboFei** [comment]
> I think the change looks reasonable. Did you check them only in `SQLQuerySuite`? I mean, have you checked if the logs of the other test suites have no `TableAlreadyExistsException`?

Hi, I found this issue when adding UT into hive/SQLQuerySuite.
There is no `TableAlreadyExistsException` in the logs before add UT.

I just checked the sql/SQLQuerySuite and refactor it.

**Comment by turboFei** [comment]
Will try to refactor other suites.

**Comment by turboFei** [comment]
I have used `grep` command to find all tests, which does clean up their used views, and refactor them.

cc @maropu

**Comment by maropu** [comment]
retest this please

**Comment by maropu** [comment]
I think `createOrReplaceTempView` doesn't throw `TableAlreadyExistsException` without `withTempView`. Does the title describe a different issue?

**Comment by turboFei** [comment]
Oh, I mean that if we createOrReplaceTempView and does not clean up it after this test.
If another test create a table with same name, `TableAlreadyExistsException` would be thrown.

**Comment by maropu** [comment]
Ur, I see. cc: @dongjoon-hyun

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/SQLQuerySuite.scala [hunk 3]
```diff
@@ -1236,43 +1273,47 @@ class SQLQuerySuite extends QueryTest with SQLTestUtils with TestHiveSingleton {
...
-    val df = sql(
-      s"""FROM test
-        |SELECT TRANSFORM(a, b)
-        |ROW FORMAT SERDE 'org.apache.hadoop.hive.serde2.lazy.LazySimpleSerDe'
-        |WITH SERDEPROPERTIES('field.delim' = '|')
-        |USING 'python $scriptFilePath/scripts/test_transform.py "|"'
-        |AS (c STRING, d STRING)
-        |ROW FORMAT SERDE 'org.apache.hadoop.hive.serde2.lazy.LazySimpleSerDe'
-        |WITH SERDEPROPERTIES('field.delim' = '|')
+    withTempView("data") {
```
Comment (maropu) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/SQLQuerySuite.scala:1297]: `date` -> `test`?

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/json/JsonSuite.scala [hunk 4]
```diff
@@ -833,171 +861,182 @@ abstract class JsonSuite extends QueryTest with SharedSparkSession with TestJson
...
-          |select complexArrayOfStruct[0].field1[1].inner2[0], complexArrayOfStruct[1].field2[0][1]
-          |from jsonTable
+      checkAnswer(
+        sql("select arrayOfStruct[0].field1, arrayOfStruct[0].field2 from jsonTable"),
+        Row(true, "str1")
+      )
+      checkAnswer(
+        sql(
+          """
+            |select complexArrayOfStruct[0].field1[1].inner2[0],
```
Comment (maropu) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/json/JsonSuite.scala:989]: nit: wrong indents.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/json/JsonSuite.scala [hunk 5]
```diff
@@ -833,171 +861,182 @@ abstract class JsonSuite extends QueryTest with SharedSparkSession with TestJson
...
+            |complexArrayOfStruct[1].field2[0][1]
+            |from jsonTable
         """.stripMargin),
```
Comment (maropu) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/json/JsonSuite.scala:992]: nit: wrong indents.

**Comment by turboFei** [comment]
thanks, would fix all indents related with `""" """`.

**Comment by turboFei** [comment]
Have refactor the indents  and double check it.
cc @maropu  @dongjoon-hyun

**Comment by maropu** [comment]
btw, since I think we tend to forget the cleanup, any other solution to automatically clean up temp views with low overheads after each test?

**Comment by turboFei** [comment]
> btw, since I think we tend to forget the cleanup, any other solution to automatically clean up temp views with low overheads after each test, if this issue's annoying?

It is difficult to identify which view we should clean up. 
For exmaple:
https://github.com/apache/spark/blob/88d39e5a89cb76286ed8201c4d8b94d8a6199a42/sql/core/src/test/scala/org/apache/spark/sql/sources/TableScanSuite.scala#L139-L151

If this temp view is created in beforeAll, we should not clean up it.

**Comment by maropu** [comment]
Ah, I see.

**Review by HyukjinKwon — COMMENTED** [review]
Seems fine to me.

**Comment by maropu** [comment]
retest this please

**Review by maropu — APPROVED** [review]
LGTM, too.

**Comment by maropu** [comment]
Thanks! Merged to master/3.0.

**Pull request closed without merging**